lib/protobuf/any.ex

```diff
@@ -0,0 +1,100 @@
+defmodule Protobuf.Any do
+  @moduledoc """
+  Provides functions for working with the `google.protobuf.Any` well known type.
+
+  Using the Any type provides the ability to send any message type in your messages (a polymorphic field).
+
+  See: [Documentation for the Any type](https://developers.google.com/protocol-buffers/docs/proto3#any)
+
+  Utilizing these functions allows you to pack and unpack any fields with any protocol buffer message.any()
+
+  Given a message type of
+
+  ```protobuf
+  message ErrorStatus {
+    string message = 1;
+    repeated google.protobuf.Any details = 2;
+  }
+  ```
+
+  Use the `Protobuf.Any.pack` function to set the information.
+
+  ```elixir
+  details = some_list_of_protobuf_messages()
+  ErrorStatus.new(%{
+    message: "There was an error",
+    details: Enum.map(details, &Protobuf.Any.pack/1)
+  })
+  ```
+
+  """
+  @type_url_prefix "type.googleapis.com/"
+  @type option :: {:prefix, module()}
+
+  @doc """
+  Pack a protocol buffer message into a `Google.Protobuf.Any` message.
+
+  This sets the correct `type_url` of the pattern: `type.googleapis.com/<package>.<message name>`
+  and the `value` of the `Any` as the serialized original message.
+
+  ## Example
+
+  ```elixir
+  Google.Protobuf.Any.new(%{
+    type_url: "type.googleapis.com/some.package.My.Message",
+    value: Some.Package.My.Message.encode(data)
+  }) = Protobuf.Any.pack(data)
+  ```
+  """
+  @spec pack(struct()) :: Google.Protobuf.Any.t()
+  def pack(%mod{} = data) do
+    Google.Protobuf.Any.new(%{
+      type_url: "#{@type_url_prefix}#{mod.full_name()}",
```

Comment:
@hassox I’m not sure this is a good name to use for this function. "Full name" doesn't tell me much about what this is. Some other possible options:

  1. `fully_qualified_name`
  2. `name_for_type_url`
  3. `name_for_any_type`

@ericmj thoughts? Other ideas?

Reply:
Updated to `fully_qualified_name`